File: Python/_1711_Count_Good_Meals/Solution.py
# coding=utf-8
# Date: 2021/7/7 9:41
from typing import List


# 执行用时：1132 ms, 在所有 Python3 提交中击败了63.31%的用户
# 内存消耗：20.4 MB, 在所有 Python3 提交中击败了62.59%的用户
class Solution:
    def countPairs(self, deliciousness: List[int]) -> int:
        n = len(deliciousness)
        mod = int(1e9) + 7

        s = [0 for _ in range(22)]
        p = 1
        for i in range(22):
            s[i] = p
            p *= 2

        res = 0
        hash = {}
        for val in deliciousness:
            for i in range(22):
                t = s[i] - val
                if t in hash:
                    res = (res + hash[t]) % mod
            if val not in hash:
                hash[val] = 0
            hash[val] += 1
        return res


if __name__ == "__main__":
    print(Solution().countPairs([1, 3, 5, 7, 9]))  # 4

File: Java/src/_1711_Count_Good_Meals/Solution.java
package _1711_Count_Good_Meals;

import java.util.HashMap;

/**
 * Date: 2021/7/7 9:37
 * Content:
 * 执行用时：171 ms, 在所有 Java 提交中击败了45.08%的用户
 * 内存消耗：49.4 MB, 在所有 Java 提交中击败了37.96%的用户
 */
public class Solution {

    public int countPairs(int[] deliciousness) {
        int n = deliciousness.length;
        int mod = (int) (1e9) + 7;

        int[] s = new int[22];
        for (int i = 0, p = 1; i < 22; i++) {
            s[i] = p;
            p *= 2;
        }

        int res = 0;
        HashMap<Integer, Integer> hash = new HashMap<>();  // (数字，出现次数)
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < 22; j++) {
                int t = s[j] - deliciousness[i];
                res = (res + hash.getOrDefault(t, 0)) % mod;
            }
            int d = deliciousness[i];
            hash.put(d, hash.getOrDefault(d, 0) + 1);
        }
        return res;
    }

    public static void main(String[] args) {

        int[] deliciousness = {1, 3, 5, 7, 9};
        System.out.println((new Solution()).countPairs(deliciousness));  // 4
    }
}
